protocols/request-response/src/codec.rs

```diff
@@ -78,3 +82,152 @@ pub trait RequestResponseCodec {
+    ) -> io::Result<()>
+    where
+        T: AsyncWrite + Unpin + Send;
+
+    /// Writes a response to the given I/O stream according to the
+    /// negotiated protocol.
+    async fn write_response<T>(
+        &mut self,
+        protocol: &Self::Protocol,
+        io: &mut T,
+        res: Self::Response,
+    ) -> io::Result<()>
+    where
+        T: AsyncWrite + Unpin + Send;
+}
+
+#[allow(deprecated)]
+impl<U> Codec for U
+where
+    U: RequestResponseCodec,
+{
+    type Protocol = U::Protocol;
+
+    type Request = U::Request;
+
+    type Response = U::Response;
+
+    fn read_request<'life0, 'life1, 'life2, 'async_trait, T>(
+        &'life0 mut self,
+        protocol: &'life1 Self::Protocol,
+        io: &'life2 mut T,
+    ) -> core::pin::Pin<
+        Box<
+            dyn core::future::Future<Output = io::Result<Self::Request>>
+                + core::marker::Send
+                + 'async_trait,
+        >,
+    >
+    where
+        T: AsyncRead + Unpin + Send,
+        T: 'async_trait,
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        Self: 'async_trait,
+    {
+        self.read_request(protocol, io)
+    }
+
+    fn read_response<'life0, 'life1, 'life2, 'async_trait, T>(
+        &'life0 mut self,
+        protocol: &'life1 Self::Protocol,
+        io: &'life2 mut T,
+    ) -> core::pin::Pin<
+        Box<
+            dyn core::future::Future<Output = io::Result<Self::Response>>
+                + core::marker::Send
+                + 'async_trait,
+        >,
+    >
+    where
+        T: AsyncRead + Unpin + Send,
+        T: 'async_trait,
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        Self: 'async_trait,
+    {
+        self.read_response(protocol, io)
+    }
+
+    fn write_request<'life0, 'life1, 'life2, 'async_trait, T>(
+        &'life0 mut self,
+        protocol: &'life1 Self::Protocol,
+        io: &'life2 mut T,
+        req: Self::Request,
+    ) -> core::pin::Pin<
+        Box<dyn core::future::Future<Output = io::Result<()>> + core::marker::Send + 'async_trait>,
+    >
+    where
+        T: AsyncWrite + Unpin + Send,
+        T: 'async_trait,
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        Self: 'async_trait,
+    {
+        self.write_request(protocol, io, req)
+    }
+
+    fn write_response<'life0, 'life1, 'life2, 'async_trait, T>(
+        &'life0 mut self,
+        protocol: &'life1 Self::Protocol,
+        io: &'life2 mut T,
+        res: Self::Response,
+    ) -> core::pin::Pin<
+        Box<dyn core::future::Future<Output = io::Result<()>> + core::marker::Send + 'async_trait>,
+    >
+    where
+        T: AsyncWrite + Unpin + Send,
+        T: 'async_trait,
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        Self: 'async_trait,
+    {
+        self.write_response(protocol, io, res)
+    }
+}
```

Comment:
Why can't we use the `async_trait` macro for this?

Reply:
we can yeah, but we need to upgrade `U`'s trait  to be `: RequestResponseCodec + Send` and `U::Protocol` to be`: Sync` due to how `#[async_trait]` expands.  
Don't know if it will break `RequestResponseCodec` impl's but updated Thomas.

Reply:
Hmm okay. I'd expect most codecs to be ZST so this shouldn't matter really. Let's see if someone complains :)